src/goto-programs/README.md

```diff
@@ -534,3 +534,79 @@ C++ modules:
 Dependencies:
   - [linking folder](\ref linking).
   - [typecheck](\ref section-goto-typecheck).
+
+\section goto-trace-structure Goto Trace Structure
+
+In the \ref goto-programs directory.
+
+**Key classes:**
+* \ref goto_tracet
+* \ref goto_trace_stept
+* \ref goto_trace_stept::typet
+
+A trace represents the execution of a program as a series of steps. The class 
+\ref goto_tracet contains an ordered list of \ref goto_trace_stept, each 
+representing a step. The main step types used in a program are:
+- function call
+- function return
+- assignment
+
+There are many types of step in goto_trace_stept, which are listed here: 
+\ref goto_trace_stept::typet, but this overview will focus on the main types 
+used.
+
+\subsection general-step-structure General Step Structure
+
+Every goto_trace_stept has a type (\ref goto_trace_stept::typet), e.g. function
+call. The type of a step can be checked using the boolean functions such as 
+\ref goto_trace_stept::is_function_call(). 
+
+The meaning of the remainder of the members of goto_trace_stept depend on the
+step type.
+
+\subsection function-call-step-structure Function Calls
+
+A function call step has the goto_trace_stept::typet 
+\ref goto_trace_stept::typet::FUNCTION_CALL 
+and signifies that the program is calling a function.
+The step contains a valid identifier of the function being called 
+(\ref goto_trace_stept::called_function), a vector of the function arguments 
+(\ref goto_trace_stept::function_arguments) and the identifier of the function
+of the call site (\ref goto_trace_stept::function).
+
+Following a function call, \ref assignment-step-structure assignments are made
+to each parameter, if there are any, including the `this` parameter.
+
+\subsection function-return-step-structure Function Return
+
+A function return step has the goto_trace_stept::typet 
+\ref goto_trace_stept::typet::FUNCTION_RETURN and signifies that the program is 
+exiting a function. \ref goto_trace_stept::function is the function being 
+returned from.
+
+If the function has a return value, an \ref assignment-step-structure
+assignment will be made to a variable with identifier with suffix
+\ref remove_returns::RETURN_VALUE_SUFFIX.
+
+\subsection assignment-step-structure Assignment
+
+An assignment step has the goto_trace_stept::typet
+\ref goto_trace_stept::typet::ASSIGNMENT and signifies that the program is
+making an assignment to a variable. For an assignment LHS = RHS, the step
+contains the symbol representing the LHS
+(\ref goto_trace_stept::full_lhs) and the value expression representing the RHS
+(\ref goto_trace_stept::full_lhs_value).
+
+
+\subsection other-steps Other steps
+
+In a trace of a program, there are other step types that are internal
+including:
+- goto_trace_stept::typet::ASSERT (failure), used in bytecode instrumentation 
+and properties
+- goto_trace_stept::typet::NONE (location-only), indicates that this step is no
```

Comment:
location-only should have `LOCATION` ?!

Reply:
I've updated this in the other PR. I wrote this on the last doc day. I've double checked and they are indeed LOCATION, not NONE, which will be changed here when rebased.